---
layout: page
title: "Notes on video of mapping RNA-seq reads"
---

```{r options, echo=FALSE}
library(knitr)

```

# Short introduction to mapping RNA-Seq reads

Note that the commands used in this lab require you have a lot of free disk space (the FASTQ files alone are 28 GB) and many cores available for running the alignment program. We do not expect students to replicate the commands in this video. We do not expect students install the alignment software on their machines. Much of the software for processing NGS data is designed for Linux systems. Note that the case studies (in particular the variant discovery and genotyping case study) will go into more depth on using Linux for processing NGS data.

The FASTQ file we are looking at in the beginning of this screencast was downloaded from:

http://trace.ncbi.nlm.nih.gov/Traces/sra/?run=SRR1177756

This is a human RNA-Seq sample from a [study](http://trace.ncbi.nlm.nih.gov/Traces/sra/?study=SRP032775) of naturally acquired immunity to malaria.

We discuss the following software in the screencast:

* fastq-dump from the [SRA toolkit](http://www.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?cmd=show&f=software&m=software&s=software)
* [Bowtie2](http://bowtie-bio.sourceforge.net/bowtie2/index.shtml)
* [Tophat2](http://ccb.jhu.edu/software/tophat/index.shtml)
* [STAR](https://github.com/alexdobin/STAR)
* [Samtools](http://samtools.sourceforge.net/)



To extract the FASTQ files from the SRA file, we used the following line. The `--split-files` argument is used to extract two files for the two paired-ends of the fragments which were sequenced.

```
fastq-dump --split-files SRR1177756.sra
```
# STAR commands


The STAR paper:

http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3530905/

The STAR manual:

https://github.com/alexdobin/STAR/blob/master/doc/STARmanual.pdf

Downloading genome FASTA and GTF files from ENSEMBL:

http://ensembl.org

http://ensembl.org/info/data/ftp/index.html

Generating the genome:

Note the `sjdbOverhang` is used for constructing the splice junction database. It should be set to (read length - 1), and according to the manual a general value of 100 will work as well.

For this limited demonstration, I am only going to align to the genes on chromosome 1, so I subset the GTF file:

```
grep -P '^1\t' Homo_sapiens.GRCh38.79.gtf > Homo_sapiens.GRCh38.79.chrom1.gtf
```

We then moved files in subdirectories, and created one for the STAR genome index:


mkdir gtf
mkdir genome
mv *.gtf gtf
mv *.fa genome
mkdir GRCh38.79.chrom1

The STAR command to generate the genome index:

STAR --runMode genomeGenerate \
--genomeDir GRCh38.79.chrom1 \
--genomeFastaFiles genome/Homo_sapiens.GRCh38.dna.chromosome.1.fa \
--sjdbGTFfile gtf/Homo_sapiens.GRCh38.79.chrom1.gtf \
--sjdbOverhang 62 


Mapping the reads:

```
STAR --runThreadN 12 \
--genomeDir GRCh38.79.chrom1 \
--readFilesIn fastq/SRR1039508_1.fastq fastq/SRR1039508_2.fastq

STAR --runThreadN 12 \
--genomeDir Homo_sapiens.GRCh38.dna.chromosome.1 \
--readFilesIn fastq/SRR1039508_1.fastq/SRR1039508_2.fastq

To filter out unmapped reads from BAM files (option -F)
samtools view -bF SRR1039508_2.fastq.bam > mappedSRR_2.bam
SRR1039508_1.fastq.bam > mappedSRR_1.bam
```
```
### SUBREAD
subread-build index 2022_GENOMICS_COURSE/Homo_sapiens.GRCh38.dna.chromosome.1.fa -o 2022_GENOMICS_COURSE/idxchr1/Homo_sapiens.GRCh38.dna.chromosome.1.fa.index

subread-align -a 2022_GENOMICS_COURSE/gtf/Homo_sapiens.GRCh38.79.gtf -B 3 -T 5 -i 2022_GENOMICS_COURSE/idxchr1/Homo_sapiens.GRCh38.dna.chromosome.1.fa.index -t 0 -r 2022_GENOMICS_COURSE/fasta/SRR1039508_1.fastq 2022_GENOMICS_COURSE/fasta/SRR1039508_2.fastq -o 2022_GENOMICS_COURSE/idxchr1/result.bam

samtools view -b 4 -F result.bam > pairmapped.bam
samtools sort -n pairmapped.bam -o pairsorted.bam



```


### TOPHAT
The call for running Tophat2 was:

```
export BOWTIE2_INDEXES=/path/to/your/Bowtie2Index

tophat2 -o SRR1177756_tophat_out -p 10 genome SRR1177756_1.fastq SRR1177756_2.fastq
```

To view the reads we used Samtools:
```
samtools view accepted_hits.bam | head -1000 | less
```

For demonstration purposes (you wouldn't necessarily repeat these lines in a typical workflow), we merged the mapped and unmapped reads into a single sorted file. For this we used the following calls:

```
samtools sort -n accepted_hits.bam accepted_hits_name_sorted
samtools sort -n unmapped.bam unmapped_name_sorted
samtools merge -n all_reads.bam accepted_hits_name_sorted.bam unmapped_name_sorted.bam
```

