---
title: "outline"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
##Chapter 0 - Introduction
Introduction [Rmd]
  1. Getting started [Rmd]
Getting started exercises
  2  dplyr introduction [Rmd]
  3. dplyr introduction exercises
  4. galton

##Chapter 1 - Inference
  5. Random variables [Rmd]
Random variables exercises
  6. CLT and t-distribution [Rmd]
CLT and t-distribution exercises
  7. CLT in practice [Rmd]
CLT in practice exercises
Confidence intervals [Rmd]
  8. Power calculations [Rmd]
Power calculations exercises
X Monte carlo [Rmd]
X Monte carlo exercises
X Permutation tests [Rmd]
X Permutation tests exercises
X Association tests [Rmd]
X Association tests exercises

##Chapter 2 - Exploratory Data Analysis
Exploratory data analysis [Rmd]
Plots to avoid [Rmd]
Exploratory data analysis exercises

##Chapter 3 - Robust Statistics
Robust summaries [Rmd]
Rank tests [Rmd]
Robust summaries exercises

##Chapter 4 - Matrix Algebra
Introduction to using regression [Rmd]
Introduction to using regression exercises
  9. Matrix notation [Rmd]
Matrix notation exercises
  10. Matrix operations [Rmd]
Matrix operations exercises
  11. Matrix algebra examples [Rmd]
Matrix algebra examples exercises

##Chapter 5 - Linear Models
Linear models introduction [Rmd]
Linear models introduction exercises
Expressing design formula [Rmd]
Expressing design formula exercises
  12. Linear models in practice [Rmd]
Linear models in practice exercises
Standard errors [Rmd]
Standard errors exercises
  13, Interactions and contrasts [Rmd]
Interactions and contrasts exercises
Collinearity [Rmd]
Collinearity exercises
QR and regression [Rmd]
Linear models going further [Rmd]

##Chapter 6 - Inference for High-Dimensional Data
  14. Introduction to high-throughput data [Rmd] -- 3 tables - add to this
Introduction to high-throughput data exercises
Inference for high-throughput data [Rmd]
Inference for high-throughput data exercises
  15. Multiple testing [Rmd]
Multiple testing exercises
  16. EDA for high-throughput data [Rmd]
EDA for high-throughput data exercises

##Chapter 7 - Statistical Modeling
Modeling [Rmd]
Modeling exercises
Bayes theorem [Rmd]
Bayes theorem exercises
Hierarchical models [Rmd]
Hierarchical models exercises

##Chapter 8 - Distance and Dimension Reduction
 17. Distance [Rmd]
Distance exercises
 18. PCA motivation [Rmd]
 19. SVD [Rmd]
SVD exercises
Projections [Rmd]
Rotations [Rmd]
 #20. MDS [Rmd]
MDS exercises
PCA [Rmd]
 20. PCA and SVD in R [Rmd]
 21. MDS [Rmd]
##Chapter 9 - Practical Machine Learning
 22. Clustering and heatmaps [Rmd]
Clustering and heatmaps exercises
Conditional expectation [Rmd]
Conditional expectation exercises
Smoothing [Rmd]
Smoothing exercises
Machine learning [Rmd]
Crossvalidation [Rmd]
Crossvalidation exercises

##Chapter 10 - Batch Effects
Introduction to batch effects [Rmd]
 23. Confounding [Rmd]
Confounding exercises
 24.. EDA with PCA [Rmd]
EDA with PCA exercises
 25. Adjusting with linear models [Rmd]
Adjusting with linear models exercises
Factor analysis [Rmd]
Factor analysis exercises
Adjusting with factor analysis [Rmd]
Adjusting with factor analysis exercises

##Chapter 11 - Introduction to Bioconductor
Mike Love’s general reference card
Motivations and core values (optional)
Installing Bioconductor and finding help [Rmd]
Data structure and management for genome scale experiments [Rmd]
Coordinating multiple tables: ExpressionSet
Institutional archives: GEO, ArrayExpress
Interlude: Working with general genomic features using GenomicRanges
IRanges introduced
Intra-range operations
Inter-range operations
GRanges
Calculating overlaps
Range-oriented solutions for current experimental paradigms
SummarizedExperiment: for RNA-seq and 450k methylation
External storage for very large assays
GenomicFiles for families of BAM or BED
DNA Variants: VCF handling with VariantAnnotation and VariantTools
Handling multiomic archives like TCGA
Cloud-oriented solutions: e.g., Google BigQuery
Short read mapping/alignment software (optional) [Rmd]

##Chapter 12 - Genomic Annotation with Bioconductor
More details on GRanges [Rmd]
Run-length encoding, views
Application to genomic landmarks
Application to 450k methylation array visualization
General overview of Bioconductor annotation [Rmd]
Levels: reference sequence, regions of interest, pathways
Discovering reference sequence
A build of the human genome
Gene/Transcript/Exon catalogs from UCSC and Ensembl
Importing and exporting regions and scores
AnnotationHub: brokering thousands of annotation resources
OrgDb: simple interface to annotation databases
Finding and managing gene sets
OrganismDb: unifying diverse annotation
Cheat sheet on Bioconductor annotation [Rmd]
Translating addresses between genome builds: liftOver [Rmd]

##Chapter 13 - Genome-scale hypothesis testing with Bioconductor
Distinguishing biological and technical variability [Rmd]
An experiment with pooled and individual samples
Measuring technical variation
Measuring biological variation
Interpretation
Multiple comparisons with genewise t-tests [Rmd]
Gene-wise testing
Naive enumeration of genes
Demonstrating danger of multiple testing with a set of sham comparisons
Adjusting for multiplicity with qvalue
Adjusted counts in the sham case
Moderated t tests via limma [Rmd]
A spike-in dataset
Naive t-tests
Three steps with limma: lmFit, eBayes, topTable
Exposing the spiked-in genes
A view of the shrinkage of variance estimates
Introducing gene sets and gene set analysis [Rmd]
Data wrangling
A dataset for comparing expression by gender
Finding surrogate variables/batch effect correction
The Broad Institute MsigDb
Identifier remapping
Categorical testing
Statistical summaries for sets: Wilcoxon
Statistical summaries for sets: t statistics
Adjusting for within-set correlation
A permutation procedure

##Chapter 14 - Visualization of genome scale data
A basic overview of visualization tasks and strategies[Rmd]
Gene models
Gene models plus data
Driving visualizations with functions
Using the browser to drive visualization functions via shiny
Queriable dynamic displays with plotly
Annotation-oriented visualizations
Sketching the binding landscape over chromosomes with ggbio’s karyogram layout [Rmd]
Plotting data in the context of genomic features with Gviz [Rmd]
Visualizing NGS data [Rmd]
Interactive visualization
Graphical user interfaces for multivariate data with shiny [Rmd]
Clustering gene expression data with shiny [Rmd]
Final remarks on visualization [Rmd]

##Chapter 15: Pursuing scalability in genomic analysis: parallelism and out-of-memory data
Parallel computing with R and Bioconductor [Rmd]
Demonstrating simple speedup in multicore environments
Implicit parallelism with BiocParallel and GenomicAlignments
External data: data interfaces that spare RAM[Rmd]
SQLite for annotation
Tabix-indexed BAM
HDF5
An illustration of NoSQL with S4: mongodb and RaggedMongoExpt[Rmd]
Benchmarking various out-of-memory solutions[Rmd]
Introduction to Bioconductor’s Amazon Machine Instance for cluster creation and use in EC2 [Rmd]
Sharded GRanges for scalable integrative analysis[Rmd]

##Chapter 16: Multi-omic data integration
Basic examples of multi-omic integration[Rmd]
Transcription factor (TF) binding and gene coexpression in yeast
TF binding and GWAS hits in humans
Using RTCGAToolbox outputs to integrate clinical, mutation, expression and methylation assays[Rmd]
Basic data acquisition
Working with clinical data
Defining a severity marker
Extracting survival times
Working with mutations
Curation tasks for discrepant identifier formats
Working with expression data
Associating tumor stage with expression patterns
Linking DNA methylation with expression patterns
Application to visualization: kataegis and rainfall plot[Rmd]

##Chapter 17: Fostering reproducible genome-scale analysis
Overview of unit on reproducibility[Rmd]
Basic definitions
Infrastructure requirements
Statistical aspects of reproducibility
Analysis of reproducibility probability (Boos and Stefanski 2011)
Costs of highly reproducible designs
Package structure, creation, installation, management[Rmd]
What is a package?
Using package.skeleton
Using makeOrganismPackage
Using devtools
  create() to set up folders and DESCRIPTION
  Composing documentation plus code
  document(), install()
Conclusions, including a link to a recent Nature Toolbox article on Bioconductor

##Legacy material from 2015 Introduction to Bioconductor
Installing Bioconductor and finding help
Annotating phenotypes and molecular function
The ExpressionSet Container

IRanges and GRanges
Operating on GRanges

Cheat sheet for genomic annotation
Translating addresses between reference builds with liftOver
Cheat sheet for GRanges and other Bioconductor objects

Importing NGS data with Bioconductor
NGS read counting

Technical versus biological variability
Statistical Inference with Bioconductor
Using limma
Gene Sets Analysis
Gene Sets Analysis in R

##RNA-seq data analysis
Downloading and unzipping fastq files
Genomic alignment with STAR
Transcriptome alignment with RSEM

RNA-seq at the gene-level: EDA, DE and SVA
Differential exon usage
Exploring plots of isoform-level expression with Cufflinks/cummeRbund

##Variant Discovery and Genotyping
Genome variation from 2014

##ChIP-seq data analysis
ChIP-seq from 2014

##DNA methylation data analysis
DNA Methylation Data Analysis
Reading 450K idat files with the minfi package
Interactive visualization of DNA methylation data analysis

Statistical Inference in the Analysis of DNA methylation Data
