---
title: "outline"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

Module 2: ‘Omics data analysis methods
    
Lectire 1  What Is Genomic Data Science?
        
•	Getting started; install R 02getting-started
•	Hands-on with R and Bioconductor packages 02biomotiv
•	Introduction to R 03dplyr-tutorial
•	Inference 05Ierence	

Lecture 2       

•        Linear regressions in R: 09matrix-notation
•        High throughput data: 14highthroughput -- 3 tables
•	 Volcano plots, histograms, boxplots, MA plots 14high lose                                            multiple testing, modeling, linear regression
•	 Dimension Reduction: 17tance	(includes clustering)


Lecture 3 Next Generation Sequencing 
      
•	NGS pipelines: Fastq, short read mapping & alignment, BAM files: 30align
•	NGS read counting, importing NGS data with Bioconductor
•	Working with general genomic features using GenomicRanges 
•	Annotation overview: 36anno
•       RNA-Seq
•       Variant Discovery and Genotyping
•       Chip-seq and data analysis
•       Epigenomics: DNA Methylation Arrays and analysis

Lec†ure 4

•       RNA-seq at the gene-level: 34rnaseq_gene_level
•       Clustering and heatmaps: Hierarchical & Kmeans moved to Dimension Reduction: 
• 	Confounding effects: Combart& SVA: 23confound

Lecture 5 GO ontology Enrichment Analysis

•	Gene Set Enrichment Analysis: 35GSEA
•	Gene Sets Analysis

3 hr	Lab Session: GO enrichment exercise
	Giaever and Nislow datasets


Module 2 Omic data analysis (9 Topics)

Module 2 Omic data analysis 
1.	Introduction to R programming: 
  a.	Getting started
  b.	DPLYR

2.	Inference
  a.	Central limit theorem
  b.  t-tests
  c.	P-values, power & confidence intervals 
  
3.  Linear Algebra
  a. Matrix notation
  b. Matrix operations
  c. Matrix algebra
  
3.	Linear regression
  a.	Solving linear models with matrices
  b.	Linear Models: Expressing design formula
  c.	Solving linear models with lm
  d.	Regressions in R
  e.  Expressing design formulas
  f.	Interactions and contrasts 

4.	Introduction to high-throughput data: 
  a.  3 tables in genomics
  b. 	Technical vs biological replicates
  c.	Multiple comparisons with genewise t-tests

Exploratory data analysis for high-throughput data
a.	Volcano plots, histograms, boxplots, MA plots
b.	Dimension Reduction: Distance, PCA, SVD, MDS

5.	Practical Machine Learning
a.	Clustering and heatmaps: Hierarchical & Kmeans
b.	Confounding effects: Combar & SVA

6.	Interlude: Working with general genomic features using GenomicRanges 
a.	IRanges & GRanges
b.	Calculating overlaps

7.	    Next Generation Sequencing Corey talk NGS
a.	Introduction to Bioconductor
b.	Installing bioconductor and motivaltion; Cheatsheet
c.	NGS pipeline:Fastq, short read mapping & alignment, BAM files
d.	NGS read counting, importing NGS data with Bioconductor
e.	Data structure and management for genome scale experiments 

CN introduces RNA-seq
a.	RNA-seq at the gene-level:
b.	EDA, DE and SVA 
c.	Differential exon usage
d.	Exploring plots of isoform-level expression
e.	scRNA-seq 
f.	CN introduces Chip-seq     
ChIP-seq data analysis

8.	CN introduces GO enrichment      
a.	Gene Set Enrichment Analysis
b.	Gene Sets Analysis in R

9.	CN introduces genotyping 
Variant Discovery and Genotyping
  
10.	CN introduces epigenomics; DNA Methylation Arrays  
a.	DNA Methylation Arrays
b.	Data analysis DNA Methylation Arrays
c.	Reading 450K idat files with the minfi package
d.	Interactive visualization of DNA methylation data analysis
e.	Inference for DNA methylation
f.	ATAC-seq 

11.	CN introduces merging datasets  
Multomic analysis
 

1.	Introduction to R programming: 
        a.	Getting started
        b.	Installing Bioconductor
        c.	DPLYR

2.	Distributions,
        a.	Central limit theorem
        b.	Association tests
        c.	P-values, power & confidence intervals 
3.	Linear regression
        a.	Matrix algebra, notation, operations 
        b.	Linear Models: Expressing design formula
        c.	Solving linear models with matrices,r
        d.	Regressions in R
        e.	Interactions and contrasts 

4.	Introduction to high-throughput data: 
                  a. 	 3 tables in genomics
                  b. 	Technical vs biologica replicates
                  c.	Multiple comparisons with genewise t-tests

5.	Exploratory data analysis for high-throughput data
        a.	Volcano plots, histograms, boxplots, MA plots
        b.	Dimension Reduction: Distance, PCA, SVD, MDS

6.	Practical Machine Learning
        a.	Clustering and heatmaps: Hierarchical & Kmeans
        b.	Confounding effects: Combar & SVA
        
        res <- t( sapply(1:nrow(geneExpression),function(j){
        y <- geneExpression[j,]
        fit <- lm(y~X-1)
        summary(fit)$coef[2,c(1,4)]
        } ) )
        res <- data.frame(res)
        names(res) <- c("dm","p.value")
        fit <- lmFit(geneExpression,X)
        k <- 2 ##second coef
        ses <- fit$stdev.unscaled[,k]*fit$sigma
        ttest <- fit$coef[,k]/ses
        pvals <- 2*pt(-abs(ttest),fit$df)
        q =qvalue::qvalue(pvals)$qvalues

        pvals = sapply(1:nrow(geneExpression),function(i){
        y = geneExpression[i,]
        fit = lm(y~X)
        summary(fit)$coef[3,4]
        })
        ###this is the 3rd coefficient!!!, the 2nd would be 2,4

        
        
7.	Next Generation Sequencing Corey talk NGS
        a.	Introduction to Bioconductor
        b.	Installing bioconductor and motivaltion; Cheatsheet
        c.	NGS pipeline:Fastq, short read mapping & alignment, BAM files
        d.	NGS read counting, importing NGS data with Bioconductor
        e.	Data structure and management for genome scale experiments 

8.	Interlude: Working with general genomic features using  GenomicRanges 
        a.	IRanges & GRanges
        b.	Calculating overlaps
    
9.	Corey talk RNA-seq
        a.	RNA-seq at the gene-level:
                b.	EDA, DE and SVA 
        c.	Differential exon usage
        d.	Exploring plots of isoform-level expression
10.	 Corey talk GO enrichment      
        a.	Gene Set Enrichment Analysis
        b.	Gene Sets Analysis in R

11.	Corey talk genotyping 
                Variant Discovery and Genotyping
  
12.	 Corey talk Chip-seq     
                ChIP-seq data analysis
  
13.	Corey talk  epigenomics; DNA Methylation Arrays  
        a.	DNA Methylation Arrays
        b.	Data analysis DNA Methylation Arrays
        c.	Reading 450K idat files with the minfi package
        d.	Interactive visualization of DNA methylation data analysis
        e.	Inference for DNA methylation

14.	Corey talk merging datasets  
                Multomic analysis


Introduction to R programming: 
  Getting started
  DPLYR

Distributions,
  Central limit theorem
  Association tests
  P-values, power & confidence intervals 

Linear algebra: Matrix Algebra, matrix notation, matrix operations 

Linear Models: Expressing design formula, solving linear models with matrices,regressions in R, standard errors, interactions and contrasts 

Introduction to high-throughput data: 
  The 3 tables in genomics
  Technical vs Biologica Replicates
  Multiple comparisons with genewise t-tests


Exploratory data analysis for high-throughput data [Rmd]
  Volcano plots, histograms, boxplots, MA plots
   Dimension Reduction
   Distance, PCA, SVD, MDS

Practical Machine Learning
  Clustering and heatmaps
  Hierarchical, KMeans
  Confounding effects: Batch effects, SVA

Corey talk NGS
Introduction to Bioconductor
  Installing bioconductor and motivaltion
  Cheatsheet
  NGS pipeline:Fastq, short read mapping & alignment, BAM files
  NGS read counting
  Importing NGS data with Bioconductor
  
  
Data structure and management for genome scale experiments 
  Coordinating multiple tables: ExpressionSet
  Institutional archives: GEO, ArrayExpress
  Annotation overview        

  
Interlude: Working with general genomic features using  GenomicRanges 
      IRanges introduced
      GRanges
      Calculating overlaps
    
Corey talk RNA-seq
RNA-seq
      RNA-seq at the gene-level: EDA, DE and SVA 
      Differential exon usage
      Exploring plots of isoform-level expression with Cufflinks/cummeRbund

Corey talk GO enrichment      
Gene Set Enrichment Analysis
      Gene Sets Analysis in R

Corey talk genotyping 
Variant Discovery and Genotyping
  
Corey talk Chip-seq     
ChIP-seq data analysis
  
Corey talk DNA Methylation Arrays  
DNA Methylation Arrays
      Data analysis DNA Methylation Arrays
      Reading 450K idat files with the minfi package
      Interactive visualization of DNA methylation data analysis
      Inference for DNA methylation

Corey talk merging datasets  
Multomic analysis

##Chapter 0 - Introduction
     Introduction [Rmd]
  1. Getting started [Rmd]
Getting started exercises
  2  dplyr introduction [Rmd]
  3. dplyr introduction exercises
  4. galton

##Chapter 1 - Inference
  5. Random variables [Rmd]
Random variables exercises
  6. CLT and t-distribution [Rmd]
CLT and t-distribution exercises
  7. CLT in practice [Rmd]
CLT in practice exercises
Confidence intervals [Rmd]
  8. Power calculations [Rmd]


##Chapter 2 - Exploratory Data Analysis
Exploratory data analysis [Rmd]
Plots to avoid [Rmd]
Exploratory data analysis exercises

##Chapter 3 - Robust Statistics
Robust summaries [Rmd]
Rank tests [Rmd]
Robust summaries exercises

##Chapter 4 - Matrix Algebra
Introduction to using regression [Rmd]
Introduction to using regression exercises
  9. Matrix notation [Rmd]
Matrix notation exercises
  10. Matrix operations [Rmd]
Matrix operations exercises
  11. Matrix algebra examples [Rmd]
Matrix algebra examples exercises

##Chapter 5 - Linear Models
Linear models introduction [Rmd]
Linear models introduction exercises
Expressing design formula [Rmd]
Expressing design formula exercises
  12. Linear models in practice [Rmd]
Linear models in practice exercises
Standard errors [Rmd]
Standard errors exercises
  13, Interactions and contrasts [Rmd]
Interactions and contrasts exercises
Collinearity [Rmd]
Collinearity exercises
QR and regression [Rmd]
Linear models going further [Rmd]

##Chapter 6 - Inference for High-Dimensional Data
  14. Introduction to high-throughput data [Rmd] -- 3 tables - add to this
  The three tables in genomics
  Technical vs Biologica Replicates
  Multiple comparisons with genewise t-tests

  15. Multiple testing [Rmd]

  16. EDA for high-throughput data [Rmd]
EDA for high-throughput data exercises

##Chapter 7 - Statistical Modeling
Modeling [Rmd]
Modeling exercises
Bayes theorem [Rmd]
Bayes theorem exercises
Hierarchical models [Rmd]
Hierarchical models exercises
 
##Chapter 8 - Distance and Dimension Reduction
 17. Distance [Rmd]
Distance exercises
 18. PCA motivation [Rmd]
 19. SVD [Rmd]
SVD exercises
Projections [Rmd]
Rotations [Rmd]
 #20. MDS [Rmd]
MDS exercises
PCA [Rmd]
 20. PCA and SVD in R [Rmd]
 21. MDS [Rmd]
##Chapter 9 - Practical Machine Learning
 22. Clustering and heatmaps [Rmd]
Clustering and heatmaps exercises
Conditional expectation [Rmd]
Conditional expectation exercises
Smoothing [Rmd]
Smoothing exercises
Machine learning [Rmd]
Crossvalidation [Rmd]
Crossvalidation exercises

##Chapter 10 - Batch Effects
Introduction to batch effects [Rmd]
 23. Confounding [Rmd]
Confounding exercises
 24.. EDA with PCA [Rmd]
EDA with PCA exercises
 25. Adjusting with linear models [Rmd]
Adjusting with linear models exercises
Factor analysis [Rmd]
Factor analysis exercises
Adjusting with factor analysis [Rmd]
Adjusting with factor analysis exercises

##Chapter 11 - Introduction to Bioconductor
  
  26. Motivations and core values (optional)
      biomotiv.Rmd
  27. Installing Bioconductor and finding help [Rmd]
  28. Mike Love’s general reference card
      bioconductor_cheatsheet.Rmd
  29. Fastq Fastqc
  30. Short read mapping/alignment software (optional) [Rmd]
      bioc1_align.Rmd
  
  31. Importing NGS data with Bioconductor
  32. NGS read counting
  
  33. Data structure and management for genome scale experiments [Rmd]: Coordinating multiple tables: ExpressionSet
Institutional archives: GEO, ArrayExpress
      dataman2019.Rmd (needs editing)
      
  
  34. Interlude: Working with general genomic features using          GenomicRanges https://www.coursera.org/learn/statistical-genomics/home/welcome
      IRanges introduced
      Intra-range operations
      Inter-range operations
      GRanges
      Calculating overlaps
      bioc1_igranges.Rmd
  35. RNA-seq
      RNA-seq at the gene-level: EDA, DE and SVA
     
      Differential exon usage
      Exploring plots of isoform-level expression with Cufflinks/cummeRbund
      
  36. Gene Set Enrichment Analysis
      Gene Sets Analysis in R
  
##Variant Discovery and Genotyping
  37. Genome variation from 2014
      
##ChIP-seq data analysis
  37. ERBS
  38. ChIP-seq from 2014
##39. Variant analysis
      
##DNA methylation data analysis
  39. DNA Methylation Data Analysis
      Reading 450K idat files with the minfi package
      Interactive visualization of DNA methylation data analysis
      Inference for DNA methylation
  40  Annotation overview
      Annotation cheatsheet
  41. Multomic analysis
  
  
  

Stat
  Genomoic Ranges: Chip-Seq
        Genes as gene ranges
        Finding the nearest gene
        Annotating genes
        Application to 450k methylation array visualization
        grangesERBSEExample.Rmd
    
      
##Chapter 12 - Genomic Annotation with Bioconductor
      More details on GRanges [Rmd]
      Run-length encoding, views
      Application to genomic landmarks
      Application to 450k methylation array visualization


  General overview of Bioconductor annotation [Rmd] Needs editing
        bioc1_annoOverview.Rmd
        Levels: reference sequence, regions of interest, pathways
        Discovering reference sequence
        A build of the human genome
        Gene/Transcript/Exon catalogs from UCSC and Ensembl
        Importing and exporting regions and scores
        AnnotationHub: brokering thousands of annotation resources
        OrgDb: simple interface to annotation databases
        Finding and managing gene sets
        OrganismDb: unifying diverse annotation

   Cheat sheet on Bioconductor annotation [Rmd]
        annoCheat.Rmd
Translating addresses between genome builds: liftOver [Rmd]

##Chapter 13 - Genome-scale hypothesis testing with Bioconductor
Distinguishing biological and technical variability [Rmd]
An experiment with pooled and individual samples
Measuring technical variation
Measuring biological variation
Interpretation
Multiple comparisons with genewise t-tests [Rmd]
Gene-wise testing
Naive enumeration of genes
Demonstrating danger of multiple testing with a set of sham comparisons
Adjusting for multiplicity with qvalue
Adjusted counts in the sham case
Moderated t tests via limma [Rmd]
A spike-in dataset
Naive t-tests
Three steps with limma: lmFit, eBayes, topTable
Exposing the spiked-in genes
A view of the shrinkage of variance estimates
Introducing gene sets and gene set analysis [Rmd]
Data wrangling
A dataset for comparing expression by gender
Finding surrogate variables/batch effect correction
The Broad Institute MsigDb
Identifier remapping
Categorical testing
Statistical summaries for sets: Wilcoxon
Statistical summaries for sets: t statistics
Adjusting for within-set correlation
A permutation procedure

##Chapter 14 - Visualization of genome scale data
A basic overview of visualization tasks and strategies[Rmd]
Gene models
Gene models plus data
Driving visualizations with functions
Using the browser to drive visualization functions via shiny
Queriable dynamic displays with plotly
Annotation-oriented visualizations
Sketching the binding landscape over chromosomes with ggbio’s karyogram layout [Rmd]
Plotting data in the context of genomic features with Gviz [Rmd]
Visualizing NGS data [Rmd]
Interactive visualization
Graphical user interfaces for multivariate data with shiny [Rmd]
Clustering gene expression data with shiny [Rmd]
Final remarks on visualization [Rmd]

##Chapter 15: Pursuing scalability in genomic analysis: parallelism and out-of-memory data
Parallel computing with R and Bioconductor [Rmd]
Demonstrating simple speedup in multicore environments
Implicit parallelism with BiocParallel and GenomicAlignments
External data: data interfaces that spare RAM[Rmd]
SQLite for annotation
Tabix-indexed BAM
HDF5
An illustration of NoSQL with S4: mongodb and RaggedMongoExpt[Rmd]
Benchmarking various out-of-memory solutions[Rmd]
Introduction to Bioconductor’s Amazon Machine Instance for cluster creation and use in EC2 [Rmd]
Sharded GRanges for scalable integrative analysis[Rmd]

##Chapter 16: Multi-omic data integration
Basic examples of multi-omic integration[Rmd]
Transcription factor (TF) binding and gene coexpression in yeast
TF binding and GWAS hits in humans
Using RTCGAToolbox outputs to integrate clinical, mutation, expression and methylation assays[Rmd]
Basic data acquisition
Working with clinical data
Defining a severity marker
Extracting survival times
Working with mutations
Curation tasks for discrepant identifier formats
Working with expression data
Associating tumor stage with expression patterns
Linking DNA methylation with expression patterns
Application to visualization: kataegis and rainfall plot[Rmd]

##Chapter 17: Fostering reproducible genome-scale analysis
Overview of unit on reproducibility[Rmd]
Basic definitions
Infrastructure requirements
Statistical aspects of reproducibility
Analysis of reproducibility probability (Boos and Stefanski 2011)
Costs of highly reproducible designs
Package structure, creation, installation, management[Rmd]
What is a package?
Using package.skeleton
Using makeOrganismPackage
Using devtools
  create() to set up folders and DESCRIPTION
  Composing documentation plus code
  document(), install()
Conclusions, including a link to a recent Nature Toolbox article on Bioconductor

##Legacy material from 2015 Introduction to Bioconductor
Installing Bioconductor and finding help
Annotating phenotypes and molecular function
The ExpressionSet Container

IRanges and GRanges
Operating on GRanges

Cheat sheet for genomic annotation
Translating addresses between reference builds with liftOver
Cheat sheet for GRanges and other Bioconductor objects

Importing NGS data with Bioconductor
NGS read counting

Technical versus biological variability
Statistical Inference with Bioconductor
Using limma

Gene Sets Analysis
Gene Sets Analysis in R

##RNA-seq data analysis
Downloading and unzipping fastq files
Genomic alignment with STAR
Transcriptome alignment with RSEM

  35. RNA-seq
      RNA-seq at the gene-level: EDA, DE and SVA
      Differential exon usage
      Exploring plots of isoform-level expression with Cufflinks/cummeRbund
      
  36. Gene Sets Analysis
      Gene Sets Analysis in R

##Variant Discovery and Genotyping
Genome variation from 2014

##ChIP-seq data analysis
  37. 
  38. ChIP-seq from 2014

##DNA methylation data analysis
  39. DNA Methylation Data Analysis
      Reading 450K idat files with the minfi package
      Interactive visualization of DNA methylation data analysis

Statistical Inference in the Analysis of DNA methylation Data
